src/Microsoft.AspNetCore.Routing.Abstractions/RouteData.cs

```diff
@@ -126,10 +126,22 @@ public RouteValueDictionary Values
         /// <returns>A <see cref="RouteDataSnapshot"/> that captures the current state.</returns>
         public RouteDataSnapshot PushState(IRouter router, RouteValueDictionary values, RouteValueDictionary dataTokens)
         {
+            // Perf: this is optimized for small list sizes, in particular to avoid overhead of a native call in
+            // Array.CopyTo inside the List(IEnumerable<T>) constructor.
+            List<IRouter> routers = null;
+            if (_routers?.Count > 0)
+            {
+                routers = new List<IRouter>();
```

Comment:
It won't make a difference for the kinds of sizes we usually ( < 4), but it can't hurt. I'll put it in.